### ISSUE TITLE
Decision on app navigation 

### ISSUE THREAD

**Issue by JayBrizle** [mgolokhov/dodroid#16]
I think the design for the app navigation should get settled on before we get too deep in features added. The 3 main patterns currently are Dashboard, Slide panels, and Tabs it seems. Any thoughts or visuals(Maxim is a great drawing artist lol) on this?

**Comment by JaeW** [comment]
Agreed.  I was heading down a rabbit hole.

**Comment by mgolokhov** [comment]
)))))
I like gmail app, google play app style (classic android user experience).
Seems like slide+tabs.

**Comment by JayBrizle** [comment]
Sounds good. I did part of an app project tutorial that used that style before. Was a long time ago now though. I think that was using a ViewPager with tabs in the layout. I am currently at a part in a good Android book that is dealing with those things so maybe I can add in some of that stuff to get a rough prototype for that sort of navigation going here soon.

**Comment by mgolokhov** [comment]
Looking forward to your push request. Meanwhile i'll also study ViewPager.

**Comment by JaeW** [comment]
Just covered ViewPager in the book Android Programming:  The Big Nerd Ranch Guide.  Really like this feature.  What about a fragment (kind of like a banner) above the ViewPager which keeps track of the number of right/wrong answers?  Points earned, etc?

**Comment by JaeW** [comment]
Are you guys familiar with Google's new <a href="https://developer.android.com/design/material/index.html">Material Design </a> specs?  Looks like Google/Android is trying to standardize app appearance to be more like iPhone for API 23 and above.  <a href="http://www.google.com/design/spec/material-design/introduction.html#">More here.</a>

**Comment by JaeW** [comment]
Guys, the Android Gmail app uses a RecyclerView.  More on that <a href="http://stackoverflow.com/questions/27855401/android-5-recyclerview-like-gmail">here</a>, <a href="http://developer.android.com/training/material/lists-cards.html">here</a>, and <a href="http://developer.android.com/reference/android/support/v7/widget/RecyclerView.html">here</a>.

Mentioned I've been working through Big Nerd Book.   Exercise/Project which implements both RecyclerView and ViewPager can be found <a href="https://github.com/JaeW/CriminalIntents.git">here</a>.

In this case the RecyclerView is placed into a fragment.  When you tap on one of the list items in the RecyclerView it takes you to that list item's data displayed in a ViewPager which can be swiped.  

I'd be interested in giving the code a go, but there are two things beyond my current skill level.  

1)  In this case the data which populates the RecyclerView is contained in a <a href=http://www.tutorialspoint.com/java/java_using_singleton.htm">singleton</a>, and I don't really understand the data structure being used in dodroid app.  

2) The RecyclerView uses ViewHolders which have their own XML layout file.    This layout file assumes that the data inserted into the inflated views is consistent.  The questions for dodroid have a varying number of answers.  I have no idea whether the layout file can be adjusted at runtime to account for this difference.

**Comment by mgolokhov** [comment]
I will check. One more [link Providing Descendant and Lateral Navigation](http://developer.android.com/training/design-navigation/descendant-lateral.html).

**Comment by JayBrizle** [comment]
I've paged through the nerd ranch book. Recycle view is a newer thing that came along with Material design, cards, etc. We can go that route then if you guys want. Just have to do a bit of reading on it. Shouldn't be too hard to implement since we aren't far into the project

**Comment by mgolokhov** [comment]
I think we should draw some sketches and graphs for navigation first.

**Comment by mgolokhov** [comment]
> I don't really understand the data structure being used in dodroid app

Janice, you can ask questions in commits, check
https://github.com/mgolokhov/dodroid/commit/a05efebcff6ffdc2507c9ffb800477c8a86b070f
We'll add more comments.

**Comment by JaeW** [comment]
The Nerd Ranch chapter on RecyclerView is not great.  Best explanation/walkthrough I've found is on the You Tube channel <a href="https://www.youtube.com/channel/UCbP2HeYGC3kfHjHLMPplZuQ">slidenerd</a>.  He has a 4-part RecyclerView tutorial - part 1 <a href="https://www.youtube.com/watch?v=Wq2o4EbM74k">here</a>.  

I haven't worked with ListView, but it seems Google/Android is pushing the ReyclerView to replace ListView, GridView, etc.  What's nice about it is the fact that the layout is determined by the RecyclerView's LayoutManager.  You can switch the layout appearance easily simply by changing the LayoutManager instead of, say, having to change a ListView to a GridView in code.  It's also very elegant in terms of memory management.

**Comment by JaeW** [comment]
Agreed - sketches of its layout would be nice.  I know a graphic designer who might be interested in contributing, but if you guys want to design the UI yourselves that's cool, too.  Just let me know if you're interested.

**Comment by JaeW** [comment]
Maxim - I did a learning bliz on git and finally got dodroid into Android Studio last night and played with it.  Will go back and study the data structure tonight.  Really interesting the way you used the ScrollView to contain question answers.  Will look at that more, too, as it addresses one of the questions I had.  AND will ask those questions in commits from now on.

**Comment by mgolokhov** [comment]
Would be nice to have a graphic designer in our team.

**Comment by mgolokhov** [comment]
Desperately need custom tags or "Other".

**Comment by JaeW** [comment]
Will contact graphic designer tomorrow.  

Maxim, I gave you edit rights on both the Google sheet and the Google Form.  My time this week, I think, is going to be super limited in terms of contribution, but please feel free to make any and all changes that would make form or data collection better.

**Comment by JaeW** [comment]
Hey Maxim - Found the "Add "other" option and added it on the form.  How many "other" boxes should we provide?  It would be nice to have a way to collect those "other" categories, validate them and then add the to a checkbox.  I'm leery of spending so much time scripting/customizing this form that it takes away from learning to program Android.  

Also agree that it would be nice to have an algorithm which flags potential duplicates.  I have no clue as to how to do this, though.

**Comment by mgolokhov** [comment]
Thanks =)

My ideal use case:
The user type a question, script searches (by nouns) possible matches and show them to the user.

To add checker is low priority. There are some add-ons, but I'm not sure that the Form can be changed at runtime. We can add javascript to validate table post factum.

**Comment by mgolokhov** [comment]
Anyway we have to re-read, modify and filter all questions manually.

**Comment by JaeW** [comment]
If the user checked the 'other' category box and then filled in a new category it would be easy for us to overlook it when checking these things manually.  We already have 18 categories.  Do you think it would be possible to add a script which, if it sensed the other category was chosen, it could copy the contents of the other text box to a separate sheet or column?  I know there's a way to pull the checkbox list from another sheet.  Not important too fix now -- just wanted to get the thought documented.

**Comment by mgolokhov** [comment]
It's possible with "Data validation...". and with script you can do anything you can imagine =)

**Comment by JaeW** [comment]
Cool.  Thanks.

Hey Maxim, do you think we should set up a list of tasks in priority order?  It may not be helpful to you two, but I have a tendency to go in too many directions and end up getting nothing done because I spin my wheels.  If you don't want to do that, would you mind letting me know where I might be able to be most helpful so I can concentrate on those things (or that one thing - either way)?

**Comment by mgolokhov** [comment]
Let's check other apps with quiz and questions. Maybe can find good ideas there.
https://play.google.com/store/apps/details?id=de.habanero.quizoid
https://play.google.com/store/apps/details?id=english.grammar.test.app

**Comment by JaeW** [comment]
I'm on it.  :-D

**Comment by JaeW** [comment]
The English Grammar app:

Don't like:  
1) I'm a native English speaker (although the British might not agree) and i didn't get 100%.  :-{   I'm going to be completely paranoid about my grammar now.
2) They wait until the end to give you feedback.  If you get answer right that understanding is neurochemically reinforced if you get immediate feedback.  
3) For dodroid, assuming user gets immediate feedback on an incorrect answer, it would be nice to have another chance to: 1) try again to answer correctly, or 2) come back to it later.  I think users should have a limited number of times to get answer right before they are automatically moved along to the next question.  
5) The extra movement when you swipe from page to page  - where the questions fall into place at different rates - is really distracting.  It doesn't add anything.
6) It allows you to choose tests by subject, but only one subject at a time.  It would be nice if you could pick several subjects which are 'shuffled' or randomized 

What I liked:
1) The splash screen explaining how it worked in a very simple way (your very fun Android image) would be perfect for that.
2) All of the answers are on the screen - no need to scroll up/down
3) The colors, white space and layout are easy to read even when the font is small - even thought it's on a small screen it feels like there's a lot of space
4) The Progress report.  Would be cool if they had a bar chart or line chart which shows progress over time
5) Love the different visual cues in the Mixed Tests screen versus the By Topic screen (Mixed Tests has circular numbered buttons, By Topic is a list).  Visually you know where you are in the app without having to read anything.
5) It's easy to get started with the app - It helps user make choices.  For instance, the first choice is Intermediate/Advanced Intermediate.  After choosing you are at a new screen with a new decision to make (Mixed Tests/By Topic).  If you choose By Topic it waits for you to choose a topic before you see the Test1, Test2 or Test3 options.  This is good - prevents user fatigue from having too many choices to make all at once (see <a href="https://www.ted.com/talks/barry_schwartz_on_the_paradox_of_choice?language=en">The Paradox of Choice</a>.
6) 7) After you finish a test you have the option to View Mistakes.  It shows you a results screen with list of right/wrong answers.  When you click on an item it takes you to that question.  That list of right/wrong answers is still available on the left-hand side as a pull-out (which is very, very cool), but it's hard to find/know what it is.  Should be more arrow looking so you know it's there and what to do with it.  

All for now.

**Comment by mgolokhov** [comment]
Nice review =)

**Comment by JaeW** [comment]
It's going to be a few days or more before I can speak with my graphic design person.  BUT, I had an idea which is a little nuts, but also just might work.  There are gobs (lots - 161 in progress right now) of design courses offered through Coursera - some specifically relating to technology.  What do you think of busting in (joining) one or more of those classes and posting to see if anyone is interested in collaborating on a real-world project?  I don't think this is against Coursera's rules/honor code, blah, blah, blah.  Thoughts?

**Comment by JaeW** [comment]
Alternatively, overall I really like the English Grammar app.  I think it is a good design and incorporates most of what we have discussed.  That could be used as a guide depending on what you guys think.

**Comment by JayBrizle** [comment]
I like some of the grammar app. We will need to break things up into categories to choose from similar to how they did, along with just having overall mixed test with different difficulty levels to choice from

**Comment by JaeW** [comment]
Like your suggestion a lot.  Have you looked at the <a href="https://docs.google.com/forms/d/12t_k5dga6CPrpOeP0kb1L9JexUK8PyRJNTRHlU9t24I/viewform">question collection form</a>?  We have added quite a few categories, but the list will expand.   I think all of our current questions are at lowest level right now.  The questions are in Google Sheet <a href="https://docs.google.com/spreadsheets/d/13bmt8pwh4x4GFTnoctxkxjKjsxDtYwwXbGS6ZEB-ik8/edit#gid=1307737117">here</a>.

**Comment by JaeW** [comment]
Looking like there are a LOT of User Interface questions.  We might need to subdivide this category.

**Comment by mgolokhov** [comment]
Grammar app is nice, but they have only one right answer and a small question text.
More [apps](https://play.google.com/store/search?q=english%20test&c=apps&docType=1&sp=CAFiDgoMZW5nbGlzaCB0ZXN0egUYAMABAooBAggB%3AS%3AANO1ljKoePo&hl=en)

**Comment by JaeW** [comment]
Grammar app also seriously sucked my battery life.

**Comment by JaeW** [comment]
Talked to the graphic designer . . . . unfortunately this isn't something she has time for at the moment.  Sorry about that.

**Comment by JaeW** [comment]
I'm really digging into Google's Material Design guidelines.  I'll try to distill them down to essentials and then post only the most important.   In the meantime I was looking for a way for the 3 of us to collaborate on wireframes.  I found one called <a href="http://creately.com/">Creately</a> which looks pretty easy to use and has Android controls in its shapes library.  Its <a href="http://creately.com/plans">free plan</a> allows for 3 collaborators and 5 public diagrams.  Would that be enough?  Check it out when you guys have a chance and let me know what you think.

**Comment by mgolokhov** [comment]
Never mind, we can do re-design latter. And our app is intended for programmers)))

**Comment by JaeW** [comment]
Sorry, are you saying ignore the material design?  Or no big deal on the graphic designer?  Or both?

**Comment by JaeW** [comment]
Totally agree we can re-design later, but do we have an agreed-upon design yet?

**Comment by mgolokhov** [comment]
Never mind about designer =)

**Comment by JaeW** [comment]
Ah . .. . . cool, cool.

**Comment by JaeW** [comment]
Also wanted to mention another tool which professional Android developers are using called <a href="http://appinventor.mit.edu/explore/">App Inventor</a>.  Yes, they are also using it to teach kids to code, but it's a super easy way, once a design is agreed upon, to mock up the look and functionality before committing the hard-core code.  Just a thought.

**Comment by mgolokhov** [comment]
Creately is cool, but let's try to do it the github may. We can play with milestones, assignee and other labels.

**Comment by mgolokhov** [comment]
We can test design\navigation feel and look here https://www.fluidui.com/

**Comment by JaeW** [comment]
Okay.  There is a very basic wireframe took within GitHub called Mockup Designer.  Haven't figured out how to work it yet.

**Comment by mgolokhov** [comment]
As GTD states sometimes low level tools are better (just notebook and pen).

**Comment by mgolokhov** [comment]
GTD - Getting Things Done

**Comment by JaeW** [comment]
Fluidui looks pretty cool.  Especially the fact that it's "Delightful to use."   

Agree with the low-level tools, but they work best when we're not all thousands of miles apart.   Just my opinion, though.

**Comment by JaeW** [comment]
LOVE GTD.   Use <a href=https://www.youtube.com/watch?v=2Y3Ybh60-2Y">GQueues with GTD</a>.  It's kinda fabulous.

**Comment by JaeW** [comment]
Anyway, here's the wireframe tool in GitHub.  
http://fatiherikli.github.io/mockup-designer/#document/783e7f84-e690-68bf-9dff-5db10637cdfb

**Comment by JaeW** [comment]
I think I spoke too soon.  Don't think it's for collaborating - think it's a project on which people are collaborating.

**Comment by mgolokhov** [comment]
new "masterpiece"

![image](https://cloud.githubusercontent.com/assets/294512/11178322/1ee3a3de-8c5d-11e5-9c17-575c07aab95e.png)

**Comment by JaeW** [comment]
Where's your Android super hero?  I miss him .  .  .

**Comment by JaeW** [comment]
So what's the video reference?  Do you want to add video tutorials/hints or something like that?

**Comment by mgolokhov** [comment]
As always in the background =)

**Comment by JaeW** [comment]
Nice!

**Comment by mgolokhov** [comment]
Link to the motivational video

**Comment by JaeW** [comment]
Oh yeah . . . how could I possibly forget about the motivational video?

**Comment by JaeW** [comment]
Okay, so you have the drawer with navigation.  Will that be a permanent item on the screen?  Or a pull-out like in the English grammar app?

**Comment by mgolokhov** [comment]
Pull-out style

**Comment by JaeW** [comment]
Nice.  I like the pull-out.  Will it contain an app-map, kind of like a website site map?  Or something else?

**Comment by mgolokhov** [comment]
Navigation as in the gmail app

**Comment by JaeW** [comment]
Makes sense.  Okay, so does the start screen have questions?  Or what does it display?  We really need to give your Droid super hero a name.  ;-)

**Comment by mgolokhov** [comment]
just do it button )))

**Comment by JaeW** [comment]
K.  So they press the button . . . where does that take them?

**Comment by mgolokhov** [comment]
![image](https://cloud.githubusercontent.com/assets/294512/11178868/abe30d62-8c60-11e5-921d-3649bde29058.png)

**Comment by mgolokhov** [comment]
questions-questions...
navigation by swipe

**Comment by JaeW** [comment]
So, when they first hit this page will it tell the user to select from the spinner tags?
What will the options/overflow menu do?
Instead of the 'Commit' button, what about one of the new floating buttons?  Like in the Gmail app - bottom right-hand side.

**Comment by JaeW** [comment]
So I totally get that this app is for programmers, but programmers are still neurologically wired the same way as the rest of the species.  Therefore, if you want this to be good - if you really want people to use it and get benefit from it I would suggest taking a look at these <a href="http://developer.android.com/design/get-started/principles.html"> design guidelines</a>.  For instance, people work best when work/learning is divided into "chunks."  So maybe give them a series of 10 questions then give them feedback.  When the get a question right it helps to give them a small reward.  Hitting milestones should trigger something fun and unexpected.  That kind of thing.

**Comment by JaeW** [comment]
Questions make you think about how you want things to work . .  . . or they just annoy the flying f*ck out of you.  Questions are helping or . . . . . not?

**Comment by mgolokhov** [comment]
Totally agree. Should be
- chunks
- progress
- category
- ....

Same stuff as in the English app.
We'll do it step by step. My primary goal is to build QuestionActivity (working horse activity), so students can play with it now.

**Comment by JaeW** [comment]
A more limited scope is good.  

So we assume anyone using it (for now) is in Part 1?  That would eliminate levels for now.  

What needs to be decided before the first Activity is built?

**Comment by JaeW** [comment]
So <a href="https://www.google.com/design/spec/patterns/navigation-drawer.html#navigation-drawer-content">this</a> is pretty cool.  If you're interested in the different options on the navigation drawer.

**Comment by JaeW** [comment]
Last question, although it should have been my first.  How can I best help?  I'm done with the class - I have time this week..

**Comment by mgolokhov** [comment]
Anyone can use (folks from Part 2 can refresh old topics).
Nothing holds us (except labs and laziness) to build this Activity =)
Nobody canceled navigation graphs.

**Comment by mgolokhov** [comment]
You can review questions, build app, ask your questions about code, draw your UI sketches...

**Comment by JaeW** [comment]
I did some work on the ViewPager, but I think it will require slight re-design.  Currently you display the questions in the main activity.  Because ViewPager container needs to be inflated from an android.support.v4.view.ViewPager view I'm pretty sure you can't put it into a generic ViewGroup.  Would you (both) be okay with putting the questions into a separate activity?

**Comment by mgolokhov** [comment]
Your fork is a bit outdated, try this https://help.github.com/articles/syncing-a-fork/

**Comment by JaeW** [comment]
My bad.  I'll check it out tonight.

**Comment by mgolokhov** [comment]
> android.support.v4.view.ViewPager view I'm pretty sure you can't put it into a generic ViewGroup

As a ViewGroup successor it should be ok to use it inside other ViewGroup, then you can find it by method findViewById()

In the latest app version I reverted to two activities=)

**Comment by mgolokhov** [comment]
By the way, in windows os you need to install python 2.7 and put it in sys PATH.
You also can comment line in the script build.gradle  
<code>
preBuild.dependsOn exportQuestions;
</code>

**Comment by mgolokhov** [comment]
> @shiraeeshi 
> What about having two buttons "commit" and "next" visible at the same time?
> If user clicks next without committing, it means that he skipped the question and will return to it later (after the last question in the set).
> If user answers wrong, he has two options: either try one more time or move to the next question.
> But if user can try one question more then one times, doesn't it break the statistics? If user doesn't know the right answer and tries to guess it, he will make his statistics worse and worse with every commit. I need to understand the statistics logic you are implementing.

**Comment by mgolokhov** [comment]
Two buttons at the same time is bad, because we loose pixels. That's why we implemented viewpager with swipe (the first version was [with buttons](https://github.com/mgolokhov/dodroid/issues/1)).
Still we can have settings for such things.
